tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -1170,65 +1169,95 @@ static bool passCursorInfoForDecl(
+      CursorSymbolInfo &SymbolInfo = Data.Symbols.emplace_back();
       DeclInfo DInfo(D, Type(), /*IsRef=*/true, /*IsDynamic=*/false,
                      ArrayRef<NominalTypeDecl *>(), Invoc);
       if (auto Err =
               fillSymbolInfo(SymbolInfo, DInfo, Info->getLoc(), AddSymbolGraph,
-                             Lang, Invoc, PreviousSnaps, Allocator)) {
+                             Lang, Invoc, PreviousSnaps, Data.Allocator)) {
         // Ignore but make sure to remove the partially-filled symbol
         llvm::handleAllErrors(std::move(Err), [](const llvm::StringError &E) {});
-        Symbols.pop_back();
+        Data.Symbols.pop_back();
       }
     }
   }
 
-  SmallVector<RefactoringInfo, 8> Refactorings;
   if (AddRefactorings) {
     Optional<RenameRefInfo> RefInfo;
     if (Info->isRef())
       RefInfo = {Info->getSourceFile(), Info->getLoc(),
                  Info->isKeywordArgument()};
-    collectAvailableRenameInfo(MainInfo.VD, RefInfo, Refactorings);
-    collectAvailableRefactoringsOtherThanRename(Info, Refactorings);
+
+    /// Adds an action to \c Data if no action with the same UID exists.
+    auto AddAction = [&Data](const RefactoringInfo &NewRefactoring) {
+      bool HasRefactoringWithSameUID = llvm::any_of(
+          Data.AvailableActions, [&](RefactoringInfo &ExistingRefactoring) {
+            return ExistingRefactoring.Kind == NewRefactoring.Kind;
+          });
+      if (HasRefactoringWithSameUID) {
+        return;
+      }
+      Data.AvailableActions.push_back(NewRefactoring);
+    };
+    if (auto Rename = collectAvailableRenameInfo(MainInfo.VD, RefInfo)) {
+      AddAction(*Rename);
+    }
+    llvm::for_each(collectAvailableRefactoringsOtherThanRename(Info),
+                   AddAction);
   }
-  Refactorings.insert(Refactorings.end(), KnownRefactoringInfo.begin(),
-                      KnownRefactoringInfo.end());
 
+  return true;
+}
+
+/// Returns true on success, false on error (and sets `Diagnostic` accordingly).
+static bool passCursorInfoForDecl(
+    ResolvedValueRefCursorInfoPtr Info, bool AddRefactorings,
+    bool AddSymbolGraph, ArrayRef<RefactoringInfo> KnownRefactoringInfo,
+    SwiftLangSupport &Lang, const CompilerInvocation &Invoc,
+    std::string &Diagnostic, ArrayRef<ImmutableTextSnapshotRef> PreviousSnaps,
+    bool DidReuseAST,
+    std::function<void(const RequestResult<CursorInfoData> &)> Receiver) {
   CursorInfoData Data;
-  Data.Symbols = llvm::makeArrayRef(Symbols);
-  Data.AvailableActions = llvm::makeArrayRef(Refactorings);
+
+  bool success =
+      addCursorInfoForDecl(Data, Info, AddRefactorings, AddSymbolGraph, Lang,
+                           Invoc, Diagnostic, PreviousSnaps);
+  if (!success) {
+    return false;
+  }
+
+  Data.AvailableActions.insert(Data.AvailableActions.end(),
+                               KnownRefactoringInfo.begin(),
+                               KnownRefactoringInfo.end());
```

Comment:
This could be `append(KnownRefactoringInfo.begin(), KnownRefactoringInfo.end())`, but I see the existing code used `insert`.